enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,239 @@
+  baremetal-operator) when the actual installation finishes. From the
+  perspective of Metal3 components, the installation is done once the ISO is
+  connected. As a result, Ironic cannot disconnect the ISO from the machine's
+  BMC, nor can it update the BMC's boot device settings properly.
+- The *live ISO* flow disables Ironic *inspection* process and has to simulate
+  it via the assisted components. This creates an unnessesary coupling between
+  the assisted service and the baremetal-operator's hardware inventory format
+  (which, in turn, is derived from the Ironic's inventory format).
+- The *standard* flow is highly [extensible][hardware managers] via the
+  [ironic-agent-image][ironic-agent] container image. This extensibility does
+  not apply to the *live ISO* flow since it does not use the Ironic agent.
+
+### Goals
+
+- ZTP follows the *standard* Metal3 flow instead of *live ISO*.
+- (i)PXE boot, firwmare and RAID settings work for ZTP without introducing
+  special API in Metal3 components.
+
+### Non-Goals
+
+- Converging two agent implementations (Ironic and assisted) into a single one.
+- Making assisted components mandatory for all bare-metal deployments.
+
+## Proposal
+
+- Building CoreOS images for the *standard* flow is handled by
+  [image-customization-controller][icc] (ICC). It will be updated to use the
+  assisted service's `InfraEnv` resources, when they are present for the given
+  `BareMetalHost`.
+- When `InfraEnv` resources are present, the generated images will contain
+  Ignition configuration for both agents. The Ironic agent will be started
+  automatically and handle coordination with Ironic and baremetal-operator.
+- [Ironic agent][ironic-agent] currently contains a custom deployment plugin
+  (*custom deploy method* in baremetal-operator terms, *custom deploy step* in
+  Ironic terms) that invokes `coreos-installer` for the image installation.
+  A new custom deploy method `install_assisted` will be added to start the
+  assisted agent at the right moment.
+- The assisted service will stop explicitly disabling *inspection* on
+  `BareMetalHost`s it manages, will start using a different process based
+  on the  new `install_assisted` deploy method and stop rebooting the machine
+  in the end of the process.
+
+### User Stories
+
+N/A (this change itself does not add new user stories)
+
+### API Extensions
+
+All required API extensions are already present.
+
+### Implementation Details/Notes/Constraints
+
+#### Image building
+
+ICC will follow different code paths depending on whether the `BareMetalHost`
+has the `infraenvs.agent-install.openshift.io` annotation linking it to an
+`InfraEnv` resource. In both cases it will generate an Ignition configuration
+to start and configure the Ironic agent. In the ZTP case, the corresponding
+`InfraEnv` will be updated with the generated Ignition configuration (merging
+it with any user-provided configuration).
```

Comment:
Right, so what I would expect to see is some way for some ZTP component to indicate to ICC that it needs to do something other than the default behavior, without anything that is tightly coupled to ZTP being baked into ICC. Is there some way to pass data from ZTP to ICC in a way that ICC can include it in the image, for example?

OR is this a step on a path to making this coordinated behavior the new default for ICC? If that's the case, then saying it in the design sets a different context for reviewing the proposed changes.

Reply:
> what I would expect to see is some way for some ZTP component to indicate to ICC that it needs to do something other than the default behavior

That's what we do. ICC uses a label on PreprovisioningImage (which is copied from a label on BareMetalHost) to decide to follow the alternative path.

> Is there some way to pass data from ZTP to ICC in a way that ICC can include it in the image, for example?

Doing this will require the data to go through baremetal-operator, which is an upstream project that is not specific to OpenShift (unlike ICC).

> OR is this a step on a path to making this coordinated behavior the new default for ICC?

No, the default behavior will still be using Ironic only and building the image ourself. I think we'll end up with 2 paths for the time being.

Reply:
> The API here is really the PreprovisioningImage CRD, and it is internal to Metal³. The BMO creates the resource and retrieves the image URL from it.
> ICC is our OpenShift-specific controller for PreprovisioningImage. As such, it is the right place to be putting logic about other OpenShift-specific features, such as ZTP. Since the CRD is defined by Metal³, the only way to provide additional information (other than having it try to call in to services that may or may not exist) is via an annotation.
> This actually works out quite well because ZTP already annotates the BMHs, and Metal³ can copy that annotation to the PreprovisioningImage to make determining when to use ZTP highly efficient.

There are 2 ways to treat an API like that: separate controllers handle separate instances (like Machine, or Ingress, or storage classes) or one controller incorporates multiple sets of logic (as is proposed here). I'm suggesting that building logic into a core OCP component to know about a layered product (ZTP) introduces an unnecessary level of coupling between the layers.

Reply:
I guess we could list as an alternative that i-c-c could just ignore all CRs with an InstallEnv label, and a separate controller (probably running in the same process as BMAC) could reconcile those ones. The complex parts of the controller code are provided by upstream Metal³ anyway, so you'd only have to write a very simple ImageProvider plugin.

I'm not sure what that buys us though. A second place to vendor the controller code just adds opportunities for it to get out of sync. i-c-c would still need to know about the layered product anyway so it could ignore the label. And if the user did apply the label erroneously as you asked about in the other thread, we'd have no opportunity to supply an error message and  instead the CR would just sit there not getting reconciled forever.

Reply:
Listing this as an alternative.